File: scripts/font_convolution.py
"""
Font Convolution Module
Applies various convolutional filters to font glyphs for feature extraction and style transfer.
"""

import numpy as np
import matplotlib.pyplot as plt
from scipy import signal
from scipy.ndimage import convolve, gaussian_filter
import os


class FontConvolution:
    """Class to perform various convolutions on font glyphs"""
    
    def __init__(self, output_dir="font_analysis_results"):
        self.output_dir = output_dir
        if not os.path.exists(output_dir):
            os.makedirs(output_dir)
    
    @staticmethod
    def create_edge_detection_kernels():
        """Create various edge detection kernels"""
        # Sobel operators
        sobel_x = np.array([[-1, 0, 1],
                           [-2, 0, 2],
                           [-1, 0, 1]])
        
        sobel_y = np.array([[-1, -2, -1],
                           [0, 0, 0],
                           [1, 2, 1]])
        
        # Prewitt operators
        prewitt_x = np.array([[-1, 0, 1],
                             [-1, 0, 1],
                             [-1, 0, 1]])
        
        prewitt_y = np.array([[-1, -1, -1],
                             [0, 0, 0],
                             [1, 1, 1]])
        
        # Laplacian
        laplacian = np.array([[0, 1, 0],
                             [1, -4, 1],
                             [0, 1, 0]])
        
        # Enhanced Laplacian
        laplacian_enhanced = np.array([[1, 1, 1],
                                      [1, -8, 1],
                                      [1, 1, 1]])
        
        return {
            'sobel_x': sobel_x,
            'sobel_y': sobel_y,
            'prewitt_x': prewitt_x,
            'prewitt_y': prewitt_y,
            'laplacian': laplacian,
            'laplacian_enhanced': laplacian_enhanced
        }
    
    @staticmethod
    def create_style_kernels():
        """Create kernels for style extraction"""
        # Sharpen
        sharpen = np.array([[0, -1, 0],
                          [-1, 5, -1],
                          [0, -1, 0]])
        
        # Emboss
        emboss = np.array([[-2, -1, 0],
                          [-1, 1, 1],
                          [0, 1, 2]])
        
        # Ridge detection
        ridge_horizontal = np.array([[-1, -1, -1],
                                    [2, 2, 2],
                                    [-1, -1, -1]])
        
        ridge_vertical = np.array([[-1, 2, -1],
                                  [-1, 2, -1],
                                  [-1, 2, -1]])
        
        # Outline detection
        outline = np.array([[-1, -1, -1],
                           [-1, 8, -1],
                           [-1, -1, -1]])
        
        return {
            'sharpen': sharpen,
            'emboss': emboss,
            'ridge_horizontal': ridge_horizontal,
            'ridge_vertical': ridge_vertical,
            'outline': outline
        }
    
    def apply_convolution(self, glyph_data, kernel, mode='constant'):
        """Apply a convolution kernel to glyph data"""
        return convolve(glyph_data, kernel, mode=mode)
    
    def extract_edge_features(self, glyph_data):
        """Extract edge features using multiple kernels"""
        kernels = self.create_edge_detection_kernels()
        features = {}
        
        for name, kernel in kernels.items():
            features[name] = self.apply_convolution(glyph_data, kernel)
        
        # Compute gradient magnitude (combining Sobel x and y)
        gradient_magnitude = np.sqrt(
            features['sobel_x']**2 + features['sobel_y']**2
        )
        features['gradient_magnitude'] = gradient_magnitude
        
        # Compute gradient direction
        gradient_direction = np.arctan2(
            features['sobel_y'], features['sobel_x']
        )
        features['gradient_direction'] = gradient_direction
        
        return features
    
    def extract_style_features(self, glyph_data):
        """Extract style features using various kernels"""
        kernels = self.create_style_kernels()
        features = {}
        
        for name, kernel in kernels.items():
            features[name] = self.apply_convolution(glyph_data, kernel)
        
        return features
    
    def apply_multi_scale_convolution(self, glyph_data, kernel, scales=[1, 2, 4]):
        """Apply convolution at multiple scales"""
        multi_scale_features = {}
        
        for scale in scales:
            # Resize kernel
            if scale > 1:
                scaled_kernel = np.kron(kernel, np.ones((scale, scale)))
            else:
                scaled_kernel = kernel
            
            # Apply convolution
            result = self.apply_convolution(glyph_data, scaled_kernel)
            multi_scale_features[f'scale_{scale}'] = result
        
        return multi_scale_features
    
    def extract_texture_features(self, glyph_data, window_size=3):
        """Extract local texture features using statistical moments"""
        features = {}
        
        # Local mean
        features['local_mean'] = convolve(
            glyph_data, 
            np.ones((window_size, window_size)) / (window_size**2)
        )
        
        # Local variance
        local_sq_mean = convolve(
            glyph_data**2,
            np.ones((window_size, window_size)) / (window_size**2)
        )
        features['local_variance'] = local_sq_mean - features['local_mean']**2
        
        # Local standard deviation
        features['local_std'] = np.sqrt(np.abs(features['local_variance']))
        
        return features
    
    def visualize_convolution_results(self, glyph_data, char='A', save_path=None):
        """Visualize all convolution results"""
        edge_features = self.extract_edge_features(glyph_data)
        style_features = self.extract_style_features(glyph_data)
        texture_features = self.extract_texture_features(glyph_data)
        
        # Create comprehensive visualization
        fig = plt.figure(figsize=(20, 12))
        
        # Original
        ax = plt.subplot(4, 5, 1)
        ax.imshow(glyph_data, cmap='gray')
        ax.set_title('Original Glyph', fontweight='bold')
        ax.axis('off')
        
        # Edge features
        plot_idx = 2
        for name in ['sobel_x', 'sobel_y', 'gradient_magnitude', 'laplacian']:
            ax = plt.subplot(4, 5, plot_idx)
            ax.imshow(edge_features[name], cmap='RdBu_r')
            ax.set_title(name.replace('_', ' ').title(), fontweight='bold')
            ax.axis('off')
            plot_idx += 1
        
        # Style features
        for name in ['sharpen', 'emboss', 'outline', 'ridge_horizontal', 'ridge_vertical']:
            ax = plt.subplot(4, 5, plot_idx)
            ax.imshow(style_features[name], cmap='viridis')
            ax.set_title(name.replace('_', ' ').title(), fontweight='bold')
            ax.axis('off')
            plot_idx += 1
        
        # Texture features
        for name in ['local_mean', 'local_variance', 'local_std']:
            ax = plt.subplot(4, 5, plot_idx)
            ax.imshow(texture_features[name], cmap='plasma')
            ax.set_title(name.replace('_', ' ').title(), fontweight='bold')
            ax.axis('off')
            plot_idx += 1
        
        # Gaussian blur (different scales)
        plot_idx += 1
        for sigma in [1, 2, 3]:
            ax = plt.subplot(4, 5, plot_idx)
            blurred = gaussian_filter(glyph_data, sigma=sigma)
            ax.imshow(blurred, cmap='gray')
            ax.set_title(f'Gaussian σ={sigma}', fontweight='bold')
            ax.axis('off')
            plot_idx += 1
        
        plt.suptitle(f'Convolution Features for Character "{char}"', 
                    fontsize=18, fontweight='bold', y=0.995)
        plt.tight_layout()
        
        if save_path:
            plt.savefig(save_path, dpi=150, bbox_inches='tight')
            print(f"Saved convolution visualization to: {save_path}")
        
        plt.close()
        
        return {
            'edge_features': edge_features,
            'style_features': style_features,
            'texture_features': texture_features
        }
    
    def extract_feature_vector(self, glyph_data):
        """Extract a comprehensive feature vector from glyph data"""
        edge_features = self.extract_edge_features(glyph_data)
        style_features = self.extract_style_features(glyph_data)
        texture_features = self.extract_texture_features(glyph_data)
        
        feature_vector = []
        
        # Statistical features from each convolution
        for features_dict in [edge_features, style_features, texture_features]:
            for name, feature_map in features_dict.items():
                # Extract statistics
                feature_vector.extend([
                    np.mean(feature_map),
                    np.std(feature_map),
                    np.min(feature_map),
                    np.max(feature_map),
                    np.median(feature_map)
                ])
        
        return np.array(feature_vector)


def test_convolution():
    """Test the convolution module with sample data"""
    print("Testing Font Convolution Module...")
    
    # Create a simple test glyph (letter-like shape)
    test_glyph = np.zeros((64, 64))
    test_glyph[10:50, 20:25] = 1  # Vertical stroke
    test_glyph[30:35, 20:45] = 1  # Horizontal stroke
    test_glyph[10:50, 40:45] = 1  # Vertical stroke
    
    # Initialize convolution module
    conv = FontConvolution(output_dir="test_conv_output")
    
    # Visualize results
    features = conv.visualize_convolution_results(
        test_glyph, 
        char='Test',
        save_path="test_conv_output/convolution_test.png"
    )
    
    # Extract feature vector
    feature_vec = conv.extract_feature_vector(test_glyph)
    print(f"\nExtracted feature vector length: {len(feature_vec)}")
    print(f"Feature vector sample: {feature_vec[:10]}")
    
    print("\nConvolution test complete!")


if __name__ == "__main__":
    test_convolution()

File: webapp/js/font-convolution.js
/**
 * font-convolution.js — Convolution filters & 80-D feature vector
 * Operates on rasterised glyph image data.
 */
window.FontConvolution = (() => {

    /* ── Kernel definitions ────────────────────────────────────── */
    const KERNELS = {
        // Edge detection
        'Sobel X':       { cat:'edge', k:[[-1,0,1],[-2,0,2],[-1,0,1]] },
        'Sobel Y':       { cat:'edge', k:[[-1,-2,-1],[0,0,0],[1,2,1]] },
        'Prewitt X':     { cat:'edge', k:[[-1,0,1],[-1,0,1],[-1,0,1]] },
        'Prewitt Y':     { cat:'edge', k:[[-1,-1,-1],[0,0,0],[1,1,1]] },
        'Roberts X':     { cat:'edge', k:[[1,0],[0,-1]] },
        'Roberts Y':     { cat:'edge', k:[[0,1],[-1,0]] },
        'Laplacian':     { cat:'edge', k:[[0,1,0],[1,-4,1],[0,1,0]] },
        'Laplacian Diag':{ cat:'edge', k:[[1,1,1],[1,-8,1],[1,1,1]] },
        'Scharr X':      { cat:'edge', k:[[-3,0,3],[-10,0,10],[-3,0,3]] },
        'Scharr Y':      { cat:'edge', k:[[-3,-10,-3],[0,0,0],[3,10,3]] },

        // Style
        'Sharpen':       { cat:'style', k:[[0,-1,0],[-1,5,-1],[0,-1,0]] },
        'Emboss':        { cat:'style', k:[[-2,-1,0],[-1,1,1],[0,1,2]] },
        'Ridge':         { cat:'style', k:[[-1,-1,-1],[-1,8,-1],[-1,-1,-1]] },
        'Unsharp Mask':  { cat:'style', k:[[1,4,6,4,1],[4,16,24,16,4],[6,24,-476,24,6],[4,16,24,16,4],[1,4,6,4,1]].map(r=>r.map(v=>v/-256)) },

        // Texture / Blur
        'Box Blur 3':    { cat:'texture', k:[[1,1,1],[1,1,1],[1,1,1]].map(r=>r.map(v=>v/9)) },
        'Box Blur 5':    { cat:'texture', k: Array(5).fill(null).map(()=>Array(5).fill(1/25)) },
        'Gaussian 3':    { cat:'texture', k:[[1,2,1],[2,4,2],[1,2,1]].map(r=>r.map(v=>v/16)) },
        'Gaussian 5':    { cat:'texture', k:[[1,4,6,4,1],[4,16,24,16,4],[6,24,36,24,6],[4,16,24,16,4],[1,4,6,4,1]].map(r=>r.map(v=>v/256)) },
        'High-Pass':     { cat:'texture', k:[[-1,-1,-1],[-1,9,-1],[-1,-1,-1]] },

        // Multi-scale
        'Gradient Mag':  { cat:'multiscale', k:null, fn:'gradientMag' },
        'LoG (5×5)':     { cat:'multiscale', k:[[0,0,1,0,0],[0,1,2,1,0],[1,2,-16,2,1],[0,1,2,1,0],[0,0,1,0,0]] },
        'Diagonal':      { cat:'multiscale', k:[[2,-1,0],[-1,2,-1],[0,-1,2]] },
    };

    /* ── Helpers ───────────────────────────────────────────────── */
    function getGrayData(canvas) {
        const ctx = canvas.getContext('2d');
        const w = canvas.width, h = canvas.height;
        const id = ctx.getImageData(0, 0, w, h);
        const gray = new Float32Array(w * h);
        for (let i = 0; i < w * h; i++) gray[i] = id.data[i * 4 + 3] / 255; // use alpha
        return { gray, w, h };
    }

    function convolve(gray, w, h, kernel) {
        const kh = kernel.length, kw = kernel[0].length;
        const oh = Math.floor(kh / 2), ow = Math.floor(kw / 2);
        const out = new Float32Array(w * h);
        for (let y = 0; y < h; y++) {
            for (let x = 0; x < w; x++) {
                let sum = 0;
                for (let ky = 0; ky < kh; ky++) {
                    for (let kx = 0; kx < kw; kx++) {
                        const ny = Math.min(h - 1, Math.max(0, y + ky - oh));
                        const nx = Math.min(w - 1, Math.max(0, x + kx - ow));
                        sum += gray[ny * w + nx] * kernel[ky][kx];
                    }
                }
                out[y * w + x] = sum;
            }
        }
        return out;
    }

    function gradientMag(gray, w, h) {
        const sx = convolve(gray, w, h, KERNELS['Sobel X'].k);
        const sy = convolve(gray, w, h, KERNELS['Sobel Y'].k);
        const out = new Float32Array(w * h);
        for (let i = 0; i < w * h; i++) out[i] = Math.sqrt(sx[i] ** 2 + sy[i] ** 2);
        return out;
    }

    function floatToCanvas(data, w, h) {
        let mn = Infinity, mx = -Infinity;
        for (const v of data) { if (v < mn) mn = v; if (v > mx) mx = v; }
        const range = mx - mn || 1;
        const c = document.createElement('canvas');
        c.width = w; c.height = h;
        const ctx = c.getContext('2d');
        const id = ctx.createImageData(w, h);
        for (let i = 0; i < w * h; i++) {
            const v = ((data[i] - mn) / range) * 255;
            id.data[i * 4 + 0] = v;
            id.data[i * 4 + 1] = v;
            id.data[i * 4 + 2] = v;
            id.data[i * 4 + 3] = 255;
        }
        ctx.putImageData(id, 0, 0);
        return c;
    }

    /* ── Run all filters ───────────────────────────────────────── */
    function runAll(sourceCanvas) {
        const { gray, w, h } = getGrayData(sourceCanvas);
        const results = {};
        for (const [name, def] of Object.entries(KERNELS)) {
            let data;
            if (def.fn === 'gradientMag') {
                data = gradientMag(gray, w, h);
            } else if (def.k) {
                data = convolve(gray, w, h, def.k);
            } else {
                continue;
            }
            results[name] = {
                canvas: floatToCanvas(data, w, h),
                data, w, h, cat: def.cat,
            };
        }
        return results;
    }

    /* ── Feature vector (statistics per filter) ────────────────── */
    function computeFeatureVector(results) {
        const features = {};
        for (const [name, r] of Object.entries(results)) {
            const d = r.data;
            let sum = 0, sum2 = 0, mn = Infinity, mx = -Infinity;
            for (const v of d) {
                sum += v; sum2 += v * v;
                if (v < mn) mn = v; if (v > mx) mx = v;
            }
            const n = d.length;
            const mean = sum / n;
            const std = Math.sqrt(sum2 / n - mean ** 2);
            features[name + ' mean'] = mean;
            features[name + ' std'] = std;
            features[name + ' min'] = mn;
            features[name + ' max'] = mx;
        }
        return features;
    }

    return { KERNELS, runAll, computeFeatureVector };
})();
